Edit profile page

PaulNewling opened the pull request:
Edit profile page that allows users to edit email, twitter and phone.
The user can also attach a github user profile. The editpage will display error text to the user if the github account does not exist.
The form for attaching a github also does not render if there is already a github attached.
Video on slack channel

PaulNewling committed:
Added edit page with updates for phone/email/twitter.
TODO: Add github logic as well as add link from Dashboard

PaulNewling committed:
Added link to edit page from dashboard.
Added link back to dashboard from Edit page.
Show error warning when github username does not exist.

PaulNewling committed:
Removed commented out code

jonabantao commented on front-end/src/api/DashboardApi.js:
```diff
@@ -9,3 +9,59 @@ export async function pullUserInfo(userID, token) {
 }
 
+//Attempts to link Github Account
+export async function linkGithubAccount(userID, token, github_username) {
```
Take the url variable and move it to the top level outside any functions. It will be evaluated once and then can be passed in through any of these API calls

jonabantao commented on front-end/src/api/DashboardApi.js:
```diff
@@ -9,3 +9,59 @@ export async function pullUserInfo(userID, token) {
+}
+
+
+// Updates Profile Information
```
@AndreEPaul @mcmoonstruck How would you feel about the frontend sending the whole user body with each PUT request to update the profile and just updating each of those fields so all these update profile functions can be consolidated down to one?

AndreEPaul replied:
If I'm understanding correctly then yeah I think that would make sense. Like you essentially do an UPDATE [every field] and any old info will just stay the same? I think thats a good idea for consolidating code.

mcmoonstruck replied:
@jonabantao, in this case, we have 3 fields that can be already updated with a single PUT request... and those 3 fields could just have 1 save button. I think it would be simpler on all sides to just send those 3 fields in a single request, since the github update path is already separate API call and would still need a separate save button?

Otherwise, I'm not sure the value of checking/parsing fields on the backend that we know can't be updated? (like user id, for example)

mcmoonstruck replied:
Unless we are proposing also combining the github username update into the same API call as the standard user update?

jonabantao replied:
Things like userId will be fine to be left out since it should be it will be in the path params and validated with our JWT middleware. We can just have the backend query always have the values of email, phone and twitter updated and the frontend has to provide us this information when updating the user profile everytime.

I don't suggest adding the github username to the same PUT request as this updates a different table.

mcmoonstruck replied:
Okay. I believe the backend code as it already is should support this...

https://github.com/cs361-spring2020-group8/github-connector/blob/staging/back-end/routes/users.js#L146

If the frontend sends a request body that contains email, twitter, and phone fields, the API will attempt to update all three fields.

PaulNewling replied:
I get a 500 error when I try to update more than one, so I split it into 3. Maybe I am doing something incorrectly? I will add a video to slack
![editprofile error](https://user-images.githubusercontent.com/7331525/81230723-dd65c980-8fa6-11ea-9f45-ca2abbebcb40.png)

jonabantao commented on front-end/src/components/EditProfile/index.js:
```diff
@@ -0,0 +1,285 @@
+    });
+  }
+
+
```
Leaving comment here just in case it's agreed to combine all edit profiles to one function

jonabantao commented on front-end/src/components/EditProfile/index.js:
```diff
@@ -0,0 +1,285 @@
+
+  handleGithubProfileFormSubmit = async (event) =>{
+    event.preventDefault();
+    let token = localStorage.getItem('access_token');
```
Prefer using const or let if the variable is not going to be changed.

65~67 and all similar code below can probably be moved out to a helper function to adhere to the DRY principle.

PaulNewling committed:
Updated edit page to return '' for twitter and phone instead of null, resolving errors thrown to console when they were empty.
Changed App title to Github Connector.

PaulNewling committed:
Combined URL check (production or local) to one line in DashboardApi.
Made token a state and combined all the userID and token fetching into one function in EditProfile

PaulNewling committed:
Edit Profile now has one button and updates all 3 forms with one function call instead of split out into 3.
Backend patch was added to user-helpers.js

PaulNewling commented:
Requests made have been completed:

- URL moved out of DashboardAPI call, duplicates deleted
- Edit Profile now only has one button, and updates all information with one function call to the backend
- Redundant token and userID declarations were moved to a helper function
- Changed let to const for items not being changed

jonabantao commented on front-end/src/App.js:
```diff
@@ -33,6 +34,14 @@ function App() {
         <Route path="/login">
           <Login />
         </Route>
+        <Route path="/editprofile" render={() => (
```
`<Login />` should probably be `<Redirect to="/login" />`.

Tested it locally and the URL doesnt change if you're not logged in. For example, remove access_token from localStorage and reload the page while in /dashboard. It shows the Login screen but still shows /dashboard up top.

jonabantao commented on front-end/src/App.js:
```diff
@@ -33,6 +34,14 @@ function App() {
           <Login />
         </Route>
+        <Route path="/editprofile" render={() => (
+          isLoggedIn() ? (
```
Super low priority but look into creating a Wrapper component to protect Routes that require you to be logged in

Jedeu commented on front-end/src/api/DashboardApi.js:
```diff
@@ -1,11 +1,42 @@
+    headers:{
+      'Accept': 'application/json',
+      'Content-Type': 'application/json',
+      'Authorization': 'Bearer '  + token }
```
Small nitpick but I'd leverage ES6 here and on line 10 to do something like this:

```
'Authorization': `Bearer ${token}`
```

Jedeu commented on front-end/src/api/DashboardApi.js:
```diff
@@ -1,11 +1,42 @@
+      'Authorization': 'Bearer '  + token }
+  };
+  const payload = {
+    'github_username': github_username
```
Another nitpick: I'd keep to the JavaScript standard and make variable names camelCase meaning:

```
{
  'github_username': githubUsername
}
```

Jedeu commented on front-end/src/components/Dashboard/index.js:
```diff
@@ -163,4 +168,9 @@ const DashboardUserContent = styled.div`
   padding: 25px 25px 25px 25px;
 `;
 
+const EditProfileLinkContainer =styled.div`
```
Nit here: add a space between `=` and `styled.div`

Jedeu commented on front-end/src/components/EditProfile/index.js:
```diff
@@ -0,0 +1,240 @@
+        }
+      } catch (e) {
+
+        const {response: {status}} = e
+        if ( status === 404) {
+          this.setState({hasGithubProfileNotFoundError: true});
+        }
+        else if (status >= 405 && status < 500) {
+          this.setState({hasAuthError: true});
+        } else {
+          this.setState({hasServerError: true});
```
At some point, I'd like to separate all of this logic into an `ApiClient` class that will call whatever endpoints and then handle any errors

Jedeu commented on front-end/src/components/EditProfile/index.js:
```diff
@@ -0,0 +1,240 @@
+          <DashboardContentContainer>
+            <DashboardUserContent>
+              {!this.state.github_info &&
+                <React.Fragment>
```
Small nit: you can use `<>` and `</>` here to make it shorter and neater.

https://reactjs.org/docs/fragments.html#short-syntax

Jedeu reviewed (commented):
I'll approve once Jon's comment is addressed

PaulNewling committed:
Added a private route wrapper to App.js
Changed the rendering of pages in App.js to redirect so the appropriate page is displayed
Updated DashboardApi.js to use ES6 for the authorization
Updated fragments to '<>' and '</>'

PaulNewling commented:
Changed the rendering of pages in App.js to redirect so the appropriate page is displayed
Updated DashboardApi.js to use ES6 for the authorization
Updated fragments to '<>' and '</>'
Minor updates that Jed pointed out were also applied

Jedeu commented on front-end/src/App.js:
```diff
@@ -8,39 +8,36 @@ import {
 import { isLoggedIn } from './helpers/auth';
 
+// Adapted from https://tylermcginnis.com/react-router-protected-routes-authentication/
+const PrivateRoute = ({ component: Component, ...rest }) => (
+  <Route {...rest} render={(props) => (
+    isLoggedIn() === true
+      ? <Component {...props} />
+      : <Redirect to='/login' />
+  )} />
+)
+
```
This is nice! I'd recommend moving it to its own file

PaulNewling committed:
privateRoute moved to its own file
